app/deepseg.cc

```diff
@@ -256,9 +265,66 @@ static bool is_number(const std::string &s) {
 	return !s.empty() && std::all_of(s.begin(), s.end(), ::isdigit);
 }
 
+std::optional<std::string> resolve_path(const std::string& provided, const std::string& type) {
+	std::string result;
+	// Check for network (URI) schema and return as-is
+	// https://www.rfc-editor.org/rfc/rfc3986#section-3.1
+	// however we require at least two chars in the scheme to allow driver letters to work on Windows..
+	if (std::regex_match(provided, std::regex("^[[:alpha:]][[:alnum:]+-.]{1,}:.*$")))
+		return provided;
+	// We use std::ifstream to check we can open each test path read-only, in order:
+	// 1. exactly what was provided
+	if (std::ifstream(provided).good())
+		return provided;
+	// to emulate PATH search behaviour (rule of least surprise), we stop here if provided has path separators
+	if (provided.find('/') != provided.npos)
+		return {};
+	// 2. BACKSCRUB_PATH prefixes if set
+	if (getenv("BACKSCRUB_PATH")!=nullptr) {
+		// getline trick: https://stackoverflow.com/questions/5167625/splitting-a-c-stdstring-using-tokens-e-g
+		std::istringstream bsp(getenv("BACKSCRUB_PATH"));
+		while (std::getline(bsp, result, ':')) {
+			result += "/" + type + "/" + provided;
+			if (std::ifstream(result).good())
+				return result;
+		}
+	}
+	// 3. XDG standard data location
+	result = getenv("XDG_DATA_HOME") ? getenv("XDG_DATA_HOME") : std::string() + getenv("HOME") + "/.local/share";
+	result += "/backscrub/" + type + "/" + provided;
+	if (std::ifstream(result).good())
+		return result;
+	// 4. prefixed with compile-time install path
+	result = std::string() + _STR(INSTALL_PREFIX) + "/share/backscrub/" + type + "/" + provided;
+	if (std::ifstream(result).good())
+		return result;
+	// 5. relative to current binary location
+	// (https://stackoverflow.com/questions/933850/how-do-i-find-the-location-of-the-executable-in-c)
+	char binloc[1024];
+	ssize_t n = readlink("/proc/self/exe", binloc, sizeof(binloc));
```

Comment:
Location of procfs mount may vary. Also not supported on some platforms.

Reply:
Yep - best option from a poor set of choices sadly, you would think this would be a trivial thing to find out..

Reply:
There's at least some existing code in [other projects](https://github.com/htop-dev/htop/blob/68c00b9cdb8ea9a841f322effb385efce1436573/configure.ac#L448-L456) \*cough\* I can point at … ;-)

Reply:
We can avoid `/proc` by using the auxiliary vector, which might be a better option: https://stackoverflow.com/a/23621200. I've been playing with this and resolving relative paths via `getcwd()` and `realpath()`, which seems pretty reliable.

Reply:
The aux vector and in particular the `AT_EXECFN` stuff seems to be glibc specials and thus likely not portable. With the `/proc` approach we have two advantages: the value can't be interfered by userland and it can be easily changed using cmake (when detecting the configuration for the system).